src/funcx_common/task_storage/s3.py

```diff
@@ -11,55 +11,100 @@
-            "key": key,
-        }
-        return True
+    def _store_to_s3(self, task: TaskProtocol, result: str) -> None:
+        try:
+            key = f"{task.task_id}.result"
+            self.client.put_object(
+                Body=result.encode("utf-8"),
+                Bucket=self.bucket_name,
+                Key=key,
+            )
+            task.result_reference = {
+                "storage_id": "s3",
+                "s3bucket": self.bucket_name,
+                "key": key,
+            }
+        except Exception:
+            raise StorageException(
+                f"Putting result into s3 for task:{task.task_id} failed"
+            )
 
-    def get_result(self, task: TaskProtocol) -> t.Optional[str]:
-        if (
-            task.result_reference
-            and task.result_reference["storage_id"] == self.storage_id
-        ):
+    def _get_from_s3(self, task: TaskProtocol) -> str:
+        assert task.result_reference
+        try:
             response = self.client.get_object(
                 Bucket=task.result_reference["s3bucket"],
                 Key=task.result_reference["key"],
             )
-
             body = response["Body"]
             return t.cast(str, body.read().decode("utf-8"))
+        except Exception:
+            raise StorageException(
+                f"Fetching object from S3 failed for: {task.task_id}"
+            )
+
+    def store_result(
+        self,
+        task: TaskProtocol,
+        result: str,
+    ) -> None:
+        if len(result) > self.redis_threshold:
+            # Task is too big for Redis, store in S3
+            self._store_to_s3(task, result)
+        else:
+            task.result = result
+            task.result_reference = {"storage_id": "redis"}
+
+        return
+
+    def get_result(self, task: TaskProtocol) -> t.Optional[str]:
+        """
+
+        :param task:
+        :return: Result if available, else raises a StorageException
+        """
+        # We should be able to safely remove the following block
+        # once all tasks launched with v0.3.3 and prior have TTL'ed out
+        if task.result:
+            return task.result
+
+        if task.result_reference:
+            if task.result_reference["storage_id"] == "s3":
+                return self._get_from_s3(task)
+
+            # The following is redundant now while the block above
+            # for backward compat exists
+            elif task.result_reference["storage_id"] == "redis":
+                return task.result
+            else:
+                raise StorageException(
+                    f"Unknown Storage requested: {task.result_reference}"
+                )
         else:
-            raise StorageException(f"Task Result was not stored with {self.storage_id}")
+            # We could return None here, because we know the result is not
+            # set at this point.
+            raise StorageException("Result appears to be not available")
```

Comment:
I would prefer that either we make this `return None` OR that the type for `get_result` be changed to `-> str` (which, by the way, becomes a pain given that `task.result` is `Optional[str]`).

As it stands, there are some cases in which you'll get `None` and some where you'll get an error, and that makes this harder to use.

Reply:
I'll update the signature to `-> str`. I'm not totally clear on if we'd call this when there's no result, or to find out whether the result is available. In my mental model, this shouldn't be called if the task isn't in a terminal state.

Reply:
Scratch that, I think it makes sense to return None when the result is not set. Since we just have one class we can definitively tell.